src/Datadog.Trace/AppSec/Agent/AppSecAgentWriter.cs

```diff
@@ -0,0 +1,103 @@
+// <copyright file="AppSecAgentWriter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Abstractions;
+using Datadog.Trace.AppSec.Transports;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec.Agent
+{
+    internal class AppSecAgentWriter : IAppSecAgentWriter
+    {
+        private const int BatchInterval = 1000;
+        private const int MaxItemsPerBatch = 1000;
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<AppSecAgentWriter>();
+        private readonly ConcurrentQueue<IEvent> _events;
+        private readonly ManualResetEventSlim _senderMutex = new(initialState: false, spinCount: 0);
+        private readonly TaskCompletionSource<bool> _processExit = new();
+        private readonly Sender _sender;
+        private readonly CancellationTokenSource _batchCancelationSource = new();
+
+        internal AppSecAgentWriter()
+        {
+            _events = new ConcurrentQueue<IEvent>();
+            Task periodicFlush = Task.Factory.StartNew(FlushTracesAsync, TaskCreationOptions.LongRunning);
+            periodicFlush.ContinueWith(t => Log.Error(t.Exception, "Error in sending appsec events"), TaskContinuationOptions.OnlyOnFaulted);
+            _sender = new Sender();
+        }
+
+        public void AddEvent(IEvent @event)
+        {
+            _events.Enqueue(@event);
+            if (!_senderMutex.IsSet)
+            {
+                _senderMutex.Set();
+            }
+        }
+
+        public void Shutdown()
+        {
+            if (!_processExit.TrySetResult(true))
+            {
+                return;
+            }
+
+            _batchCancelationSource.Cancel();
+            if (!_senderMutex.IsSet)
+            {
+                _senderMutex.Set();
+            }
+        }
+
+        private async Task FlushTracesAsync()
+        {
+            while (true)
+            {
+                if (_processExit.Task.IsCompleted)
+                {
+                    return;
+                }
+
+                if (_events.Count == 0)
+                {
+                    _senderMutex.Wait();
+                    _senderMutex.Reset();
+                }
+
+                try
+                {
+                    var appsecEvents = new List<IEvent>();
+                    while (_events.TryDequeue(out var result) && appsecEvents.Count <= MaxItemsPerBatch)
```

Comment:
In the scenario that `appsecEvents.Count > MaxItemsPerBatch` and there are remaining items in `events`, the dequeue operation will successfully remove the first item but will not send it, and therefore discard the result. Should the size check be done before accessing the queue? Or do you intend to drop events when the batch limit is exceeded?

Reply:
fixed in 56eff08dc8899e65b62d0b8786e3b281851090b4